File: programmers/level2/java/택배_배달과_수거하기.java
package programmers.level2.java;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        return findDistance(deliveries, pickups, cap) * 2;
    }

    private long findDistance(int[] deliveries, int[] pickups, int cap) {
        long distance = 0;

        int d = 0;
        int p = 0;
        for (int i = deliveries.length - 1; i >= 0; i--) {
            d += deliveries[i];
            p += pickups[i];
            while (d > 0 || p > 0) {
                distance += i + 1;
                d -= cap;
                p -= cap;
            }
        }
        return distance;
    }
}

File: programmers/level2/python/택배_배달과_수거하기.py
# 40분


def solution(cap, n, deliveries, pickups):
    distance = 0

    d = 0
    p = 0
    for i in range(n - 1, -1, -1):
        # 이번에 처리해야하는 양
        d += deliveries[i]
        p += pickups[i]

        while d > 0 or p > 0:
            distance += i + 1
            # cap만큼 처리함
            # 음수가 되면, 다음 차례에서 처리해야하는 양을 더할 때, 감안됨
            d -= cap
            p -= cap

    return distance * 2
